internal/crypto/cms/signed.go

```diff
@@ -0,0 +1,259 @@
+	// find signer certificate
+	cert := d.getCertificate(signerInfo.SignerIdentifier)
+	if cert == nil {
+		return nil, ErrCertificateNotFound
+	}
+
+	// verify signer certificate
+	certChains, err := cert.Verify(*opts)
+	if err != nil {
+		return cert, VerificationError{Detail: err}
+	}
+
+	// verify signature
+	if err := d.verifySignature(signerInfo, cert); err != nil {
+		return nil, err
+	}
+
+	// verify attribute
+	return cert, d.verifyAttributes(signerInfo, certChains)
+}
+
+// verifySignature verifies the signature with a trusted certificate.
+//
+// References:
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifySignature(signerInfo *SignerInfo, cert *x509.Certificate) error {
+	// verify signature
+	algorithm := oid.ToSignatureAlgorithm(
+		signerInfo.DigestAlgorithm.Algorithm,
+		signerInfo.SignatureAlgorithm.Algorithm,
+	)
+	if algorithm == x509.UnknownSignatureAlgorithm {
+		return VerificationError{Message: "unknown signature algorithm"}
+	}
+
+	signed := d.Content
+	if len(signerInfo.SignedAttributes) > 0 {
+		encoded, err := asn1.MarshalWithParams(signerInfo.SignedAttributes, "set")
+		if err != nil {
+			return VerificationError{Message: "invalid signed attributes", Detail: err}
+		}
+		signed = encoded
+	}
+
+	if err := cert.CheckSignature(algorithm, signed, signerInfo.Signature); err != nil {
+		return VerificationError{Detail: err}
+	}
+	return nil
+}
+
+// verifyAttributes verifies the signed attributes.
+//
+// References:
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifyAttributes(signerInfo *SignerInfo, chians [][]*x509.Certificate) error {
+	// verify attributes if present
+	if len(signerInfo.SignedAttributes) == 0 {
+		return nil
+	}
+
+	var contentType asn1.ObjectIdentifier
+	if err := signerInfo.SignedAttributes.TryGet(oid.ContentType, &contentType); err != nil {
+		return VerificationError{Message: "invalid content type", Detail: err}
+	}
+	if !d.ContentType.Equal(contentType) {
+		return VerificationError{Message: fmt.Sprintf("mismatch content type: found %q in signer info, and %q in signed data", contentType, d.ContentType)}
+	}
+
+	var expectedDigest []byte
+	if err := signerInfo.SignedAttributes.TryGet(oid.MessageDigest, &expectedDigest); err != nil {
+		return VerificationError{Message: "invalid message digest", Detail: err}
+	}
+	hash, ok := oid.ToHash(signerInfo.DigestAlgorithm.Algorithm)
+	if !ok {
+		return VerificationError{Message: "unsupported digest algorithm"}
+	}
+	actualDigest, err := hashutil.ComputeHash(hash, d.Content)
+	if err != nil {
+		return VerificationError{Message: "hash failure", Detail: err}
+	}
+	if !bytes.Equal(expectedDigest, actualDigest) {
+		return VerificationError{Message: fmt.Sprintf("mismatch message digest: found %x in signer attributes, and %x in signed data", expectedDigest, actualDigest)}
```

Comment:
We could hide the message digest when there's a mismatch. Showing the actual digest to the user does not provide more helpful information to them.
```suggestion
		return VerificationError{Message: "mismatch message digest"}
```

Reply:
Updated.